docs/designs/keras_callback.md

```diff
@@ -0,0 +1,301 @@
+# Design to Support Custom Callback Using Keras API
+
+This document describes the design for supporting callback to customize the behavior of model during training, evaluation and inference in ElasticDL.
+
+## Motivation
+
+In deep learning, we generally need to customize the behavior of model during training, evaluation or
+inference, including reading/changing the model. We may perform the behavior per batch, per epoch or at the start and end of job. `tf.keras.callbacks.Callback` is an abstract base class and has methods to perform the behavior at different call frequency, such as `on_bath_end`, `on_epoch_end` and so on. So, we adopt the interfaces of `tf.keras.callbacks.Callback` to customize the behavior of model in ElasticDL. 
+
+Now we have implemented some modules similar to callback in ElasticDL, such as `LearningRateScheduler` and `PredictionOutputsProcessor`. And users should write definitions in the model definition file for each module like:
+```python
+def custom_model():
+    ...
+
+# Adjust the learning rate according to iteration steps
+def learning_rate_scheduler(model_version):
+    if model_version < 5000:
+        return 0.0003
+    elif model_version < 12000:
+        return 0.0002
+    else:
+        return 0.0001
+
+# Process the prediction outputs for each batch
+class PredictionOutputsProcessor(BasePredictionOutputsProcessor):
+    ...
+```
+There will be multiple definition APIs for users to define different behaviors of model. The interfaces of APIs may be different. It is more convenient for users to define those behaviors using `tf.keras.callbacks.Callback`. 
```

Comment:
The interfaces of APIs may be different => The interface may be different.

API is `Application programming interface`